src/index.ts

```diff
@@ -0,0 +1,326 @@
+   */
+  saveToSerializedData(): object {
+    return {
+      recentlyUsedBlocks: this.recentlyUsedBlocks,
+      defaultJsonForBlockLookup: this.defaultJsonForBlockLookup,
+    };
+  }
+
+  /**
+   * Loads the state of this object from a serialized JSON.
+   *
+   * @param state the serialized data payload to load from
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  loadFromSerializedData(state: any): void {
+    if (state.recentlyUsedBlocks) {
+      this.recentlyUsedBlocks = state.recentlyUsedBlocks;
+    }
+    if (state.defaultJsonForBlockLookup) {
+      this.defaultJsonForBlockLookup = state.defaultJsonForBlockLookup;
+    }
+  }
+
+  /**
+   * Resets the internal state of this object.
+   */
+  clearPriorBlockData(): void {
+    this.recentlyUsedBlocks = [];
+    this.defaultJsonForBlockLookup = {};
+  }
+}
+
+/**
+ * Main entry point to initialize the suggested blocks categories.
+ *
+ * @param workspace the workspace to load into
+ * @param numBlocksPerCategory how many blocks should be included per
+ * category. Defaults to 10.
+ * @param waitForFinishedLoading whether to wait until we hear the
+ * FINISHED_LOADING event before responding to BLOCK_CREATE events. Set to false
+ * if you disable events during initial load. Defaults to true.
+ */
+export const init = function (
+  workspace: Blockly.WorkspaceSvg,
+  numBlocksPerCategory: 10,
+  waitForFinishedLoading: true,
+): void {
+  const suggestor = new BlockSuggestor(numBlocksPerCategory);
+  workspace.registerToolboxCategoryCallback(
+    'AI_SUGGESTED',
+    suggestor.getSuggestedBlocks,
+  );
+  workspace.registerToolboxCategoryCallback('MOST_USED', suggestor.getMostUsed);
+  workspace.registerToolboxCategoryCallback(
+    'RECENTLY_USED',
+    suggestor.getRecentlyUsed,
+  );
+  // If user says not to wait to hear FINISHED_LOADING event,
+  // then always respond to BLOCK_CREATE events.
+  if (!waitForFinishedLoading) suggestor.workspaceHasFinishedLoading = true;
+  workspace.addChangeListener(suggestor.eventListener);
+  suggestorLookup.set(workspace, suggestor);
+};
+
+/**
+ * Custom serializer so that the block suggestor can save and later recall which
+ * blocks have been used in a workspace.
+ */
+class BlockSuggestorSerializer implements Blockly.serialization.ISerializer {
+  /**
+   * The priority for deserializing block suggestion data.
+   * Should be less than the priority for blocks so that this state is
+   * applied after the blocks are loaded.
+   *
+   * @type {number}
+   */
+  priority = Blockly.serialization.priorities.BLOCKS - 10;
+
+  /**
+   * Saves a target workspace's state to serialized JSON.
+   *
+   * @param workspace the workspace to save
+   * @returns the serialized JSON if present
+   */
+  save(workspace: Blockly.Workspace): object | null {
+    const suggestor = suggestorLookup.get(workspace);
+    return suggestor ? suggestor.saveToSerializedData() : null;
+  }
```

Comment:
This should be how it was in the original file. If you specify the `return` type to be `object | undefined` instead of `object | null`, the compiler shouldn't produce any errors.
```suggestion
    return suggestorLookup.get(workspace)?.saveToSerializedData();
```